lib/widgets/swipable_message_row.dart

```diff
@@ -0,0 +1,115 @@
+  @override
+  Widget build(BuildContext context) {
+    final hasMarker = widget.message.editState != MessageEditState.none;
+
+    return Row(
+      crossAxisAlignment: CrossAxisAlignment.baseline,
+      textBaseline: localizedTextBaseline(context),
+      children: [
+        hasMarker
+          ? _EditStateMarker(editState: widget.message.editState)
+          : const SizedBox(width: 16),
+        ...widget.children,
+      ],
+    );
+  }
+}
+
+class _EditStateMarker extends StatelessWidget {
+  /// The minimum width of the marker.
+  ///
+  /// Currently, only the collapsed state of the marker has been implemented,
+  /// where only the marker icon, not the marker text, is visible.
+  static const double widthCollapsed = 16;
+
+  const _EditStateMarker({
+    required MessageEditState editState,
+  }) : _editState = editState;
+
+  final MessageEditState _editState;
+
+  @override
+  Widget build(BuildContext context) {
+    final theme = DesignVariables.of(context);
+    final zulipLocalizations = ZulipLocalizations.of(context);
+
+    final IconData icon;
+    final double iconSize;
+    final String markerText;
+
+    switch (_editState) {
+      case MessageEditState.none:
+        return const SizedBox(width: widthCollapsed);
+      case MessageEditState.edited:
+        icon = ZulipIcons.edited;
+        iconSize = 14;
+        markerText = zulipLocalizations.messageIsEdited;
+        break;
+      case MessageEditState.moved:
+        icon = ZulipIcons.message_moved;
+        iconSize = 8;
+        markerText = zulipLocalizations.messageIsMoved;
+        break;
+    }
+
+    var marker = Row(
+      mainAxisAlignment: MainAxisAlignment.end,
+      mainAxisSize: MainAxisSize.min,
+      children: [
+        Flexible(
+          fit: FlexFit.loose,
+          // For now, [markerText] is not displayed because it is transparent and
+          // there is not enough space in the parent ConstrainedBox.
+          child: Text('$markerText ',
+            overflow: TextOverflow.clip,
+            softWrap: false,
+            textAlign: TextAlign.center,
+            style: TextStyle(fontSize: 15, color: theme.textMarker.withAlpha(0)))),
+        // To match the Figma design, we cannot make the collapsed width of the
+        // marker larger. We need to explicitly allow the icon to overflow.
+        OverflowBox(
+          fit: OverflowBoxFit.deferToChild,
+          maxWidth: 10,
+          child: Icon(icon, size: iconSize, color: theme.textMarkerLight),
+        ),
+      ],
+    );
+
+    return ConstrainedBox(
+      constraints: const BoxConstraints(maxWidth: widthCollapsed),
+      child: Container(
+        clipBehavior: Clip.hardEdge,
+        decoration: BoxDecoration(
+          borderRadius: BorderRadius.circular(2),
+          color: theme.bgMarker.withAlpha(0)),
+        child: Padding(
+          padding: const EdgeInsets.all(1.0),
+          child: marker),
```

Comment:
I don't see vertical padding in the Figma. Am I missing it?

The height of the rounded-rectangle gets set by the line of text (e.g. "Edited"), it looks like. That's 22, because the line-height is set to that, and I don't see anything creating more vertical whitespace, in or out of the rounded-rectangle.

I do see some horizontal padding, but I wonder if this is the right way to express it. Here's a screenshot of the Figma:

<img width="983" alt="image" src="https://github.com/zulip/zulip-flutter/assets/22248748/d75026d9-e69c-4431-88a0-1d155f34d019">

In collapsed state, I see a 3px horizontal gap between the icon and the message content.

In uncollapsed state, I see:

a) 4px between the rounded-rectangle and the message content
b) 2px between the icon and the right edge of the rounded-rectangle
c) 5px between "Edited" and the icon
d) 3px between the left edge of the rounded-rectangle and "Edited"

To my eye, (a) should maybe just become 3px, to match the 3px in the collapsed state. (Maybe helps keep the implementation simple too?) And it seems to me like (b) and (d) should equal each other, to be symmetrical; not sure which value is better. Then since the uncollapse action adds space between the icon and the message content, perhaps (b) should animate from 0 to 2 as part of the animation (and then we animate (d), too, to be symmetrical).

Also, let's do (c) explicitly with a 5px gap, instead of a space character as in this revision. That might be something to animate from zero too, I'm not sure.

Reply:
Thanks for the breakdown on the design. Really helpful! I'm moving the relevant updates for this to #765.